src/Microsoft.ML.Transforms/TensorflowTransform.cs

```diff
@@ -0,0 +1,665 @@
+            }
+            return srcTensorGetters;
+        }
+
+        private unsafe T[] FetchData<T>(IntPtr data, int size)
+        {
+            var result = new T[size];
+
+            GCHandle handle = GCHandle.Alloc(result, GCHandleType.Pinned);
+            IntPtr target = handle.AddrOfPinnedObject();
+
+            Int64 sizeInBytes = size * Marshal.SizeOf((typeof(T)));
+            Buffer.MemoryCopy(data.ToPointer(), target.ToPointer(), sizeInBytes, sizeInBytes);
+            handle.Free();
+            return result;
+        }
+
+        private Delegate MakeGetter(IRow input, int iinfo)
+        {
+            var info = _bindings.Infos[iinfo];
+            var outInfo = _bindings.OutputCols[iinfo];
+            var tfType = _bindings.OutputTFTypes[iinfo];
+
+            if (tfType == TFDataType.Float)
+            {
+                return Utils.MarshalInvoke(MakeGetter<float>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            if (tfType == TFDataType.Double)
+            {
+                return Utils.MarshalInvoke(MakeGetter<double>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            if (tfType == TFDataType.Int32)
+            {
+                return Utils.MarshalInvoke(MakeGetter<int>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            if (tfType == TFDataType.Int64)
+            {
+                return Utils.MarshalInvoke(MakeGetter<Int64>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            if (tfType == TFDataType.Bool)
+            {
+                return Utils.MarshalInvoke(MakeGetter<bool>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            if (tfType == TFDataType.String)
+            {
+                return Utils.MarshalInvoke(MakeGetter<string>, outInfo.ItemType.RawType, input, iinfo);
+            }
+
+            throw Host.ExceptNotSupp("Tensorflow type not supported");
+        }
+
+        private ValueGetter<VBuffer<T>> MakeGetter<T>(IRow input, int iinfo)
+        {
+            Host.AssertValue(input);
+            // Need to check type here.
+            //Host.Assert(Infos[iinfo].TypeSrc.IsText);
+
+            var info = _bindings.Infos[iinfo];
+            var tfInfo = _bindings.TfColInfo[iinfo];
+            var srcTensorGetters = GetTensorValueGetters(input, iinfo);
+
+            return (ref VBuffer<T> dst) =>
+                {
+                    var runner = _session.GetRunner();
+                    for (int i = 0; i < info.SrcIndices.Length; i++)
+                    {
+                        var inputName = tfInfo.InputColNames[i];
+                        var type = info.SrcTypes[i];
+                        runner.AddInput(inputName, srcTensorGetters[i].GetTensor());
+                    }
+
+                    var tensors = runner.Fetch(_bindings.OutputColNames[iinfo]).Run();
```

Comment:
>tensors [](start = 24, length = 7)

Does TF know how to output values from multiple layers in one call? If this is the case, we should not make this transform ManyToOne, it should be many-to-many.

Reply:
Yes, we get multiple outputs from in one call.  I am not sure right now how to do it in ML.Net. This is the next thing I am going to focus on.

---
In reply to: [210687722](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210687722) [](ancestors = 210687722)